### PR TITLE
Evaluate text-2-sql capability using Spider benchmark

### PR DESCRIPTION
**Pull request by ekzhu**
We want to benchmark LlamaIndex's performance for complex queries on multiple domains, and measure how each iteration of LLM improves its Text-to-SQL capability, thus this PR.

### CODE REVIEW

**Comment by ekzhu** [comment]
Let me know if we can include the evaluation result as well.

File: benchmarks/struct_indices/spider/generate_sql.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+        databases[db_name] = (SQLDatabase(engine=engine), engine)
+
+    # Create the LlamaIndexes for all databases.
+    llm_predictor = LLMPredictor(
+        llm=OpenAI(temperature=0, model_name="text-davinci-003")
+    )
+    llm_indexes = {}
+    for db_name, (db, engine) in databases.items():
+        # Get the name of the first table in the database.
+        # This is a hack to get a table name for the index, which can use any
```
Comment (jerryjliu) [inline benchmarks/struct_indices/spider/generate_sql.py:112]: yeah, i'll fix this soon

File: benchmarks/struct_indices/spider/generate_sql.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    parser.add_argument(
+        "--input", type=str, required=True, help="Path to the spider dataset directory."
+    )
+    parser.add_argument(
+        "--output",
+        type=str,
+        required=True,
+        help="Path to the output directory of generated SQL files,"
+        " one query on each line, "
+        "to be compared wit the *_gold.sql files in the input directory.",
```
Comment (jerryjliu) [inline benchmarks/struct_indices/spider/generate_sql.py:79]: ubernit: with

**Review by jerryjliu — COMMENTED** [review]
seems like this is just raw text-to-SQL, would be cool if you also used some of the advanced functionality like injecting context, storing table context in this guide: https://gpt-index.readthedocs.io/en/latest/guides/sql_guide.html

of course, that can also be a separate post!

File: benchmarks/struct_indices/spider/generate_sql.py (hunk 2 above)
Comment (ekzhu) [reply benchmarks/struct_indices/spider/generate_sql.py:79]: > seems like this is just raw text-to-SQL, would be cool if you also used some of the advanced functionality like injecting context, storing table context in this guide: https://gpt-index.readthedocs.io/en/latest/guides/sql_guide.html
> 
> of course, that can also be a separate post!

Yep. Definitely.

**Comment by ekzhu** [comment]
Just tried using context container following the [documentation](https://gpt-index.readthedocs.io/en/latest/guides/sql_guide.html), the training set execution accuracy decreased from 0.407 to 0.267; the development set execution accuracy decreased from 0.5 to 0.3. 

I built the containers using schema information -- there isn't any other information available in the benchmark databases. 

The accuracy decrease is probably due to the loss of information when going from raw table schema to embedding vectors. Each Spider database has only 2 - 5 tables -- very small. Given this, I think the Text-2-SQL performance may suffer if there are many more tables, unless there is some external context information that is very relevant to the expected queries.

**Comment by jerryjliu** [comment]
> Just tried using context container following the [documentation](https://gpt-index.readthedocs.io/en/latest/guides/sql_guide.html), the training set execution accuracy decreased from 0.407 to 0.267; the development set execution accuracy decreased from 0.5 to 0.3.

oh very interesting. did you try just setting the raw context text, or also storing the context in an index?

**Comment by ekzhu** [comment]
> > Just tried using context container following the [documentation](https://gpt-index.readthedocs.io/en/latest/guides/sql_guide.html), the training set execution accuracy decreased from 0.407 to 0.267; the development set execution accuracy decreased from 0.5 to 0.3.
> 
> oh very interesting. did you try just setting the raw context text, or also storing the context in an index?

I used a separate index for the context, see [this code](https://github.com/ekzhu/gpt_index/blob/context-container/benchmarks/struct_indices/spider/generate_sql.py). 

The interesting thing is that if you go down to the underlying Langchain class, you can see it actually [includes the database schema (create table statements)](https://github.com/hwchase17/langchain/blob/master/langchain/sql_database.py#L92) in the prompts. So the LLM already has the full context available. Any additional context added to the prompt, if not bringing external information, may actually lead to redundancy and confuse the LLM.

**Comment by ekzhu** [comment]
Should we merge this request, or is there something else you would like me to add? A couple of possibilities:
1. Update the evaluation script so results yielding the same/equivalent NL response should be considered correct.
2. Make it into a Github action so we can automatically run the benchmark whenever we change code in structured index. 

Although maybe these can be separate PRs.

**Comment by jerryjliu** [comment]
> Should we merge this request, or is there something else you would like me to add? A couple of possibilities:
> 
> 1. Update the evaluation script so results yielding the same/equivalent NL response should be considered correct.
> 2. Make it into a Github action so we can automatically run the benchmark whenever we change code in structured index.
> 
> Although maybe these can be separate PRs.

Sorry missed this! yes sure, would love to land

**Review by jerryjliu — APPROVED** [review]
